Rewrite `non_exhaustive_omitted_patterns` as a separate pass

Nadrieril opened the pull request:
This is a rework of the `non_exhaustive_omitted_patterns` lint. It changes the semantics of the lint for two reasons:
- the previous implementation was inconsistent and hard to make consistent without impacting performance of exhaustiveness checking in general (discussed [here](https://github.com/rust-lang/rust/issues/89554#issuecomment-1548582793));
- this matches more closely the intent of the lint as I understand it, which is to help users keep their code up-to-date with upstream changes.

The new lint ignores considerations of exhaustiveness. Instead it looks at any part of the match that mentions some variants of a non_exhaustive enum, and triggers the lint if the match doesn't mention all the variants. That way users can ensure they update their code to mention all variants.
I also fixed some other inconsistencies in how the lint was implemented.

This separate pass is probably slow (but optional). I wouldn't recommend enabling the lint crate-wide. That said, correctness comes before performance, and I have plans that should make it much better.

Nadrieril committed:
Add some tests

Nadrieril committed:
Don't use per-arm lint levels

This makes things unnecessarily complicated for now

Nadrieril committed:
Implement lint as a separate columns-wise pass

Nadrieril commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit d5edcb14374f0a50fbec0fc5c7c42001815d2693 with merge 390ac690ed0d597a9d018264c02b97ecad184aa3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d5edcb14374f0a50fbec0fc5c7c42001815d2693","merge_sha":"390ac690ed0d597a9d018264c02b97ecad184aa3"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4995546069/jobs/8949935941)
Build commit: 390ac690ed0d597a9d018264c02b97ecad184aa3 (`390ac690ed0d597a9d018264c02b97ecad184aa3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4995546069/jobs/8949935941"},"merge_sha":"390ac690ed0d597a9d018264c02b97ecad184aa3"} -->

rust-timer commented:
Queued 390ac690ed0d597a9d018264c02b97ecad184aa3 with parent 9239760da89cd8e9a51536beeb0a23762064037d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9239760da89cd8e9a51536beeb0a23762064037d&end=390ac690ed0d597a9d018264c02b97ecad184aa3).
<!-- rust-timer: temporary -->

eholk commented:
I'm starting to look over this now, but let me first ask to make sure I understand the intent of the lint. The idea is that you want to make sure you're matching against all currently known variants of an enum, right? In other words, `#[non_exhaustive]` means you must have a wildcard, while this lint makes sure that you are still doing an exhaustive match, even if it wouldn't be a compilation error if more variants were added in the future. Is that the idea?

rust-timer commented:
Finished benchmarking commit (390ac690ed0d597a9d018264c02b97ecad184aa3): [comparison URL](https://perf.rust-lang.org/compare.html?start=9239760da89cd8e9a51536beeb0a23762064037d&end=390ac690ed0d597a9d018264c02b97ecad184aa3&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=9239760da89cd8e9a51536beeb0a23762064037d&end=390ac690ed0d597a9d018264c02b97ecad184aa3&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.2%, 0.3%] | 5     |
| Regressions ❌ <br /> (secondary)  | 3.0% | [2.8%, 3.2%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.2%, 0.3%] | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9239760da89cd8e9a51536beeb0a23762064037d&end=390ac690ed0d597a9d018264c02b97ecad184aa3&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.9% | [0.9%, 0.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9239760da89cd8e9a51536beeb0a23762064037d&end=390ac690ed0d597a9d018264c02b97ecad184aa3&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=9239760da89cd8e9a51536beeb0a23762064037d&end=390ac690ed0d597a9d018264c02b97ecad184aa3&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 642.522s -> 642.237s (-0.04%)

eholk reviewed (commented):
Still working on this, but I need to take a break so I wanted to post the couple of comments I have so far.

Nadrieril commented on compiler/rustc_lint_defs/src/builtin.rs:
```diff
@@ -3814,9 +3816,9 @@ declare_lint! {
+    /// #[warn(non_exhaustive_omitted_patterns)]
     /// match Bar::A {
     ///     Bar::A => {},
-    ///     #[warn(non_exhaustive_omitted_patterns)]
```
Not anymore, that made everything more complicated. The annotation must be on something that contains the whole match now

Nadrieril commented:
Yeah I think you got the idea. It's basically a "please remind me if someone adds a variant to this enum because I want to handle all of them".

Very annoyed at the perf result, how can it be worse I just removed code and added a single top-level check :'( Will have to investigate

eholk commented on compiler/rustc_mir_build/src/thir/pattern/usefulness.rs:
```diff
@@ -642,14 +642,7 @@ impl<'p, 'tcx> Usefulness<'p, 'tcx> {
-                                        .0
-                                        .iter()
-                                        .chain(once(pat))
-                                        .map(DeconstructedPat::clone_and_forget_reachability)
```
Why is it okay not to forget reachability now? Is there a risk of having stale or incomplete reachability information?

Nadrieril replied:
Nah this changes nothing, it's just that I hadn't implemented Clone to make sure no one cloned stuff by mistake. But it's inconvenient and there's not sensible situation where one would get confused. I have plans to clear that up more in the future

eholk commented on tests/ui/rfc-2008-non-exhaustive/omitted-patterns.rs:
```diff
@@ -31,122 +32,115 @@ pub enum Bar {
     }
 
-    #[deny(non_exhaustive_omitted_patterns)]
+    // Don't lint if no variant is mentioned, because we can't do it consistently.
```
Interesting! Why not?

Nadrieril replied:
Take for example
```rust
match Some(&NonExhaustiveEnum::A) {
  _ => {}
}
```
In order to lint here I'd need to know that the enum exists at all, which means I'd need to dig into the `_` pattern, before even knowing if there might be an interesting enum somewhere deep. I can't just dig into all wildcards like that though, that would destroy performance.

Another approach would be to somehow compute ahead of time for all types whether they have some non_exhaustive enum deep down, but that doesn't seem worth the effort.

eholk reviewed (approved):
This looks pretty good to me! I'll hold off on sending it to bors so you can chase the performance issues if you want.

I did have a couple questions that I added.

Nadrieril commented:
Alright so I'm kinda rewriting the exhaustiveness algorithm big time over in https://github.com/rust-lang/rust/pull/111720 x). It's a rewrite I've been itching to do for years but could never get right. If it works this time, the performance argument I stated in the OP is moot. In fact what I'm doing here is a lot slower because it's a whole separate pass.

So hum I don't know what to do anymore. There's still a discussion to be had about what we want the lint to do but I don't think this PR is the right solution anymore. I'll close this for now and maybe come back to it when I'm done with the other one. Thanks for taking the time to review!

Nadrieril commented:
I revived this in https://github.com/rust-lang/rust/pull/116734
